flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -461,8 +462,9 @@ public MetricTrackingOutput(
         public void emitRecord(StreamRecord<OUT> streamRecord) throws Exception {
             output.emitRecord(streamRecord);
             this.sourceMetricGroup.recordEmitted();
-            if (streamRecord.hasTimestamp()) {
-                this.sourceMetricGroup.eventTimeEmitted(streamRecord.getTimestamp());
+            long timestamp = streamRecord.getTimestamp();
+            if (timestamp != TimestampAssigner.NO_TIMESTAMP) {
```

Comment:
Can we keep the `streamRecord.hasTimestamp()` check? Now it implies that `TimestampAssigner.NO_TIMESTAMP` is always the same value as the return of `streamRecord.getTimestamp()` if no timestamp is configured.

Reply:
It doesn't make sense to keep the check as `SourceOutputWithWatermarks#collect` will never set the flag.

But I created https://issues.apache.org/jira/browse/FLINK-23858 to clarify it further and potentially deprecate the old uses.

Reply:
Thanks for opening the follow-up although the discussion goes already deeper than expected...